File: sample-videochat-webrtc/sample-videochat-webrtc/Components/Call/CallHelper/MediaController.m
//
//  MediaController.m
//  sample-videochat-webrtc
//
//  Created by Injoit on 18.10.2021.
//  Copyright © 2021 QuickBlox Team. All rights reserved.
//

#import "MediaController.h"
#import "Settings.h"
#import "VideoFormat.h"
#import "SharingScreenCapture.h"

@interface MediaController ()

@property (nonatomic, strong) QBRTCCameraCapture *camera;
@property (nonatomic, strong) SharingScreenCapture *sharing;

@property (nullable, nonatomic, strong) id appInactiveStateObserver;
@property (nullable, nonatomic, strong) id appActiveStateObserver;
@property (nonatomic, assign) BOOL didDiactivatedOutside;

@end

@implementation MediaController
@synthesize currentAudioOutput = _currentAudioOutput;

- (void)clear {
    NSNotificationCenter *defaultCenter = [NSNotificationCenter defaultCenter];
    if (self.appActiveStateObserver) {
        [defaultCenter removeObserver:(self.appActiveStateObserver)];
    }
    if (self.appInactiveStateObserver) {
        [defaultCenter removeObserver:(self.appInactiveStateObserver)];
    }
    self.appInactiveStateObserver = nil;
    self.appActiveStateObserver = nil;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _videoEnabled = NO;
        _audioEnabled = YES;
        _sharingEnabled = NO;
        _didDiactivatedOutside = NO;
        
        __weak __typeof(self)weakSelf = self;
        NSNotificationCenter *center = NSNotificationCenter.defaultCenter;
        self.appActiveStateObserver = [center addObserverForName:UIApplicationDidBecomeActiveNotification
                                                          object:nil
                                                           queue:NSOperationQueue.mainQueue
                                                      usingBlock:^(NSNotification * _Nonnull note) {
            if (!weakSelf.didDiactivatedOutside) {
                return;
            }
            weakSelf.didDiactivatedOutside = NO;
            if ([weakSelf.delegate respondsToSelector:@selector(mediaController:videoBroadcastEnable:capture:)]) {
                [weakSelf.delegate mediaController:weakSelf videoBroadcastEnable:YES capture:nil];
            }
        }];
        
        self.appInactiveStateObserver = [center addObserverForName:UIApplicationWillResignActiveNotification
                                                            object:nil
                                                             queue:NSOperationQueue.mainQueue
                                                        usingBlock:^(NSNotification * _Nonnull note) {
            if (!self.videoEnabled && !self.sharingEnabled) {
                return;
            }
            if ([weakSelf.delegate respondsToSelector:@selector(mediaController:videoBroadcastEnable:capture:)]) {
                [weakSelf.delegate mediaController:weakSelf videoBroadcastEnable:NO capture:nil];
            }
            weakSelf.didDiactivatedOutside = YES;
        }];
    }
    return self;
}

- (void)setAudioEnabled:(BOOL)audioEnabled {
    _audioEnabled = audioEnabled;
    if ([self.delegate respondsToSelector:@selector(mediaController:audioBroadcastEnable:reason:)]) {
        [self.delegate mediaController:self audioBroadcastEnable:audioEnabled reason:ChangeAudioStateReasonActionUser];
    }
}


- (void)setVideoEnabled:(BOOL)videoEnabled {
    _videoEnabled = videoEnabled;
    if (videoEnabled == YES && self.camera == nil) {
        Settings *settings = [[Settings alloc] init];
        self.camera = [[QBRTCCameraCapture alloc] initWithVideoFormat:settings.videoFormat
                                                             position:settings.preferredCameraPostion];
    }
    if ([self.delegate respondsToSelector:@selector(mediaController:videoBroadcastEnable:capture:)]) {
        [self.delegate mediaController:self videoBroadcastEnable:videoEnabled capture:self.camera];
    }
}

- (void)setSharingEnabled:(BOOL)sharingEnabled {
    _sharingEnabled = sharingEnabled;
    if (sharingEnabled == YES) {
        self.sharing = [[SharingScreenCapture alloc] initWithVideoFormat:self.videoFormat];
        if ([self.delegate respondsToSelector:@selector(mediaController:videoBroadcastEnable:capture:)]) {
            [self.delegate mediaController:self videoBroadcastEnable:YES capture:self.sharing];
        }
        return;
    }
    if ([self.delegate respondsToSelector:@selector(mediaController:videoBroadcastEnable:capture:)]) {
        [self.delegate mediaController:self videoBroadcastEnable:self.videoEnabled capture:self.camera];
    }
    self.sharing = nil;
}

- (void)sendScreenContent:(CVPixelBufferRef)content {
    
    QBRTCVideoFrame *videoFrame =
    [[QBRTCVideoFrame alloc] initWithPixelBuffer:content
                                   videoRotation:QBRTCVideoRotation_0];
    [self.sharing sendVideoFrame:videoFrame];
}

- (void)setCurrentAudioOutput:(AVAudioSessionPortOverride)currentAudioOutput {
    QBRTCAudioSession *audioSession = [QBRTCAudioSession instance];
    if (!audioSession.isActive) {
        return;
    }
    _currentAudioOutput = currentAudioOutput;
    [audioSession overrideOutputAudioPort:currentAudioOutput];
}

- (AVAudioSessionPortOverride)currentAudioOutput {
    QBRTCAudioSession *audioSession = [QBRTCAudioSession instance];
    if (!audioSession.isActive) {return AVAudioSessionPortOverrideNone;}
    NSArray<AVAudioSessionPortDescription *> *outputs = [AVAudioSession.sharedInstance currentRoute].outputs;
    for (AVAudioSessionPortDescription *output in outputs) {
        return output.portType == AVAudioSessionPortBuiltInSpeaker ?
        AVAudioSessionPortOverrideSpeaker : AVAudioSessionPortOverrideNone;
    }
    return AVAudioSessionPortOverrideNone;
}

- (QBRTCVideoTrack * _Nullable)videoTrackForUserID:(NSNumber *)userID {
    return [self.delegate mediaController:self videoTrackForUserID:userID];
}

// MARK: - CallKitManagerActionDelegate
- (void)callKit:(CallKitManager *)callKit didTapMute:(BOOL)isMuted {
    if (self.audioEnabled == !isMuted) {
        return;
    }
    _audioEnabled = !isMuted;
    if ([self.delegate respondsToSelector:@selector(mediaController:audioBroadcastEnable:reason:)]) {
        [self.delegate mediaController:self audioBroadcastEnable:!isMuted reason:ChangeAudioStateReasonActionCallKit];
    }
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/Calls/CallHelper/MediaController.swift
//
//  MediaController.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 07.12.2021.
//  Copyright © 2021 QuickBlox. All rights reserved.
//

import UIKit
import QuickbloxWebRTC

enum MediaType {
    case audio, video, sharing
}

enum ChangeAudioStateAction: UInt {
    case user, callKit
}

protocol MediaControllerDelegate: AnyObject {
    func mediaController(_ mediaController: MediaController, videoBroadcast enable: Bool, capture: QBRTCVideoCapture?)
    func mediaController(_ mediaController: MediaController, audioBroadcast enable: Bool, action: ChangeAudioStateAction)
    func mediaController(_ mediaController: MediaController, videoTrackForUserID userID: UInt) -> QBRTCVideoTrack?
}

class MediaController: NSObject {
    //MARK: - Properties
    weak var delegate: MediaControllerDelegate?
    var camera: QBRTCCameraCapture? = nil
    var sharing: SharingScreenCapture? = nil
    
    var videoFormat: VideoFormat!
    var sharingFormat: VideoFormat!
    
    private var _audioEnabled = true
    
    var currentAudioOutput: AVAudioSession.PortOverride {
        let audioSession = QBRTCAudioSession.instance()
        if audioSession.isActive {
            let outputs = AVAudioSession.sharedInstance().currentRoute.outputs
            for output in outputs {
                switch output.portType {
                case .builtInSpeaker: return .speaker
                default: return .none
                }
            }
        }
        return .none
    }
    
    private var appActiveStateObserver: NSObjectProtocol!
    private var appInactiveStateObserver: NSObjectProtocol!
    private var necessaryToContinueVideoBroadcast = false

    //MARK: - Life Cycle
    override init() {
        super.init()
        
        let center = NotificationCenter.default
        appActiveStateObserver = center.addObserver(forName: UIApplication.didBecomeActiveNotification,
                                                    object: nil,
                                                    queue: OperationQueue.main,
                                                    using: { [weak self] (note) in
            guard let self = self else {
                return
            }
            if self.necessaryToContinueVideoBroadcast == false {
                return
            }
            self.necessaryToContinueVideoBroadcast = false
            self.delegate?.mediaController(self, videoBroadcast: true, capture: nil)
        })
        
        appInactiveStateObserver = center.addObserver(forName: UIApplication.willResignActiveNotification,
                                                      object: nil,
                                                      queue: OperationQueue.main,
                                                      using: { [weak self] (note) in
            guard let self = self else {
                return
            }
            if self.videoEnabled == false, self.sharingEnabled == false {
                return
            }
            self.delegate?.mediaController(self, videoBroadcast: false, capture: nil)
            self.necessaryToContinueVideoBroadcast = true
        })
    }
    
    /// Video broadcast
    var videoEnabled = false {
        didSet {
            if videoEnabled == true, camera == nil {
                let settings = Settings()
                camera = QBRTCCameraCapture(videoFormat: settings.videoFormat,
                                            position: settings.preferredCameraPostion)
                
            }
            delegate?.mediaController(self, videoBroadcast: videoEnabled, capture: camera)
            
        }
    }
    
    /// Audio broadcast
    var audioEnabled: Bool {
        get {
            return self._audioEnabled
        }
        set {
            self._audioEnabled = newValue
            
            delegate?.mediaController(self, audioBroadcast: newValue, action: .user)
        }
    }
    
    /// Sharing broadcast
    var sharingEnabled = false {
        didSet {
            if sharingEnabled == true {
                sharing = SharingScreenCapture(sharingFormat: sharingFormat)
                delegate?.mediaController(self, videoBroadcast: true, capture: sharing)
                return
            }
            delegate?.mediaController(self, videoBroadcast: videoEnabled, capture: camera)
            sharing = nil
        }
    }
    
    func sendScreenContent(_ content: CVPixelBuffer) {
        guard let sharing = sharing else {
            return
        }
        
        let videoFrame = QBRTCVideoFrame(pixelBuffer: content, videoRotation: QBRTCVideoRotation._0)
        sharing.send(videoFrame)
    }
    
    func videoTrack(for userId: UInt) -> QBRTCVideoTrack? {
        return delegate?.mediaController(self, videoTrackForUserID: userId)
    }
}

extension MediaController: CallKitManagerActionDelegate {
    func callKit(_ callKit: CallKitManager, didTapMute isMuted: Bool) {
        if audioEnabled == !isMuted {
            return
        }
        self._audioEnabled = !isMuted
        
        delegate?.mediaController(self, audioBroadcast: !isMuted, action: .callKit)
    }
}
